File: CSharp_SetColumns/XamlBindingPathVisitor.cs
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Reflection;
using Shared;

namespace CSharp_SetColumns {
    public class XamlBindingPathVisitor : ExpressionVisitor {
        private static readonly HashSet<MethodInfo> stringFormats = typeof(string).GetMethods()
            .Where(x => {
                if (x.Name != "Format") { return false; }
                var parameters = x.GetParameters();
                return parameters.Length == 2 && parameters.First().ParameterType == typeof(string);
            })
            .ToHashSet();

        private bool isInitialized;
        private (string path, string stringFormat) current;
        public (string Path, string StringFormat) Last;

        private bool isBindablePath = true;

        public override Expression Visit(Expression node) {
            if (node is LambdaExpression lexpr) { return Visit(lexpr.Body); }

            // https://stackoverflow.com/a/34794447/111794
            bool firstVisit = !isInitialized;
            if (firstVisit) {
                current.path = "";
                isInitialized = true;
            }

            // A MethodCallExpression representing a static method call will have its Expression property set to null
            // Also, String.Format may be taking a ConstantExpression
            if (node != null && !(node is ConstantExpression)) { 
                isBindablePath =
                    node is MemberExpression ||
                    node is ParameterExpression || 
                    (node is UnaryExpression && node.NodeType.In(ExpressionType.Convert, ExpressionType.ConvertChecked));
                // TODO how can we separate between any ParameterExpression and the ParameterExpression defined by the current lambda?

                if (firstVisit) {
                    isBindablePath = isBindablePath ||
                        (node is MethodCallExpression mcexpr && stringFormats.Contains(mcexpr.Method));
                }
            }

            // https://stackoverflow.com/questions/36737463/stop-traversal-with-expressionvisitor
            if (!isBindablePath && !firstVisit) { return node; }

            var ret = base.Visit(node);

            if (firstVisit) {
                if (isBindablePath) {
                    Last = (current.path.Substring(1), current.stringFormat);
                } else {
                    Last = (null, null);
                }
                current = (null, null);
                isBindablePath = true;
                isInitialized = false;
            }
            return ret;
        }

        protected override Expression VisitMember(MemberExpression node) {
            var ret = base.VisitMember(node); ;
            current.path += "." + node.Member.Name;
            return ret;
        }

        protected override Expression VisitMethodCall(MethodCallExpression node) {
            var ret = base.VisitMethodCall(node);

            // We're assuming this is a call to string.Format; all others should be excluded in the Visit override

            if (isBindablePath && node.Arguments.Count==2 && node.Arguments[0] is ConstantExpression cexpr) {
                current.stringFormat = cexpr.Value as string;
            } else {
                isBindablePath = false;
            }

            return ret;
        }
    }
}

File: VisualBasic_SetColumns/XamlBindingPathVisitor.vb
Imports System.Reflection
Imports System.Linq.Expressions

Class XamlBindingPathVisitor
    Inherits ExpressionVisitor

    Shared ReadOnly StringFormats As HashSet(Of MethodInfo) = GetType(String).GetMethods().
        Where(Function(x)
                  If x.Name <> "Format" Then Return False
                  Dim parameters = x.GetParameters
                  Return parameters.Length = 2 AndAlso parameters.First.ParameterType = GetType(String)
              End Function).
        ToHashSet

    Private isInitialized As Boolean
    Private current As (Path As String, StringFormat As String)
    Public Last As (Path As String, StringFormat As String)

    Private isBindablePath As Boolean = True

    Public Overrides Function Visit(node As Expression) As Expression
        Dim ret As Expression = Nothing
        If ExecIfType(Of LambdaExpression)(node, Function(lexpr)
                                                     ret = Visit(lexpr.Body)
                                                 End Function) Then Return ret

        ' https://stackoverflow.com/a/34794447/111794
        Dim firstVisit = Not isInitialized
        If (firstVisit) Then
            current.Path = ""
            isInitialized = True
        End If

        ' A MethodCallExpression representing a static method call will have its Expression property set to null
        ' Also, String.Format may be taking a ConstantExpression
        If node IsNot Nothing AndAlso TypeOf node IsNot ConstantExpression Then
            isBindablePath =
                TypeOf node Is MemberExpression OrElse
                TypeOf node Is ParameterExpression OrElse
                (TypeOf node Is UnaryExpression AndAlso node.NodeType.In(ExpressionType.Convert, ExpressionType.ConvertChecked))
            ' TODO how can we separate between any ParameterExpression and the ParameterExpression defined by the current lambda?

            If firstVisit Then
                isBindablePath = isBindablePath OrElse
                    ExecIfType(Of MethodCallExpression)(node, Function(mcexpr) StringFormats.Contains(mcexpr.Method))
            End If
        End If

        ' https://stackoverflow.com/questions/36737463/stop-traversal-with-expressionvisitor
        If Not (isBindablePath OrElse firstVisit) Then Return node

        ret = MyBase.Visit(node)

        If firstVisit Then
            If isBindablePath Then
                Last = (current.Path.Substring(1), current.StringFormat)
            Else
                Last = (Nothing, Nothing)
            End If
            current = (Nothing, Nothing)
            isBindablePath = True
            isInitialized = False
        End If

        Return ret
    End Function

    Protected Overrides Function VisitMember(node As MemberExpression) As Expression
        Dim ret = MyBase.VisitMember(node)
        current.Path += "." + node.Member.Name
        Return ret
    End Function

    Protected Overrides Function VisitMethodCall(node As MethodCallExpression) As Expression
        Dim ret = MyBase.VisitMethodCall(node)

        ' We're assuming this is a call to String.Format; all others should be excluded in the Visit override

        isBindablePath = isBindablePath AndAlso node.Arguments.Count = 2 AndAlso ExecIfType(Of ConstantExpression)(node.Arguments(0), Function(cexpr)
                                                                                                                                          current.StringFormat = CStr(cexpr.Value)
                                                                                                                                      End Function)

        Return ret
    End Function

End Class
